### PR TITLE
Change mempool to save transactions in storage. [ECR-895]

### PR DESCRIPTION
**Pull request by vldm**
Save transactions in persistent pool instead of in memory `Map`. This allows node to recover safe after restart.

### CODE REVIEW

File: exonum/benches/block.rs [hunk 1]
```diff
@@ -23,41 +23,62 @@ extern crate test;
...
+
+            fn state_hash(&self, _: &Snapshot) -> Vec<Hash> {
+                Vec::new()
+            }
+
+            fn tx_from_raw(&self, raw: RawTransaction) -> Result<Box<Transaction>, EncodingError> {
+                Ok(Box::new(Tx::from_raw(raw)?))
+            }
+        }
+        messages! {
```
Comment (vitvakatu) [inline exonum/benches/block.rs]: Use `transactions!` macro?

File: exonum/src/blockchain/tests.rs [hunk 2]
```diff
@@ -50,7 +92,7 @@ fn test_encode_decode() {
 fn test_u64() {
     encoding_struct! {
         struct Test {
-            some_test: u64,
+            some_test:u64
```
Comment (vitvakatu) [inline exonum/src/blockchain/tests.rs]: What a weird change

File: sandbox/src/sandbox.rs [hunk 3]
```diff
@@ -460,34 +460,54 @@ impl Sandbox {
             .collect()
     }
 
-    /// Extracts state_hash from the fake block.
+    /// Extract state_hash from fake block
```
Comment (vitvakatu) [inline sandbox/src/sandbox.rs]: Revert this change, please

File: exonum/src/api/public/system.rs [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 The Exonum Team
+// Copyright 2017 The Exonum Team
```
Comment (ozkriff) [inline exonum/src/api/public/system.rs]: 2018

File: exonum/src/blockchain/mod.rs [hunk 5]
```diff
@@ -338,6 +334,65 @@ impl Blockchain {
...
+        tx_hash: Hash,
+        height: Height,
+        index: usize,
+        fork: &mut Fork,
+    ) -> Result<(), ::failure::Error> {
+        let tx = {
+            let schema = Schema::new(&fork);
+
+            let tx = schema.transactions().get(&tx_hash).ok_or(format_err!(
+                "BUG: Cannot find transaction in database."
```
Comment (DarkEld3r) [inline exonum/src/blockchain/mod.rs]: Please remove the dot and "BUG: " prefix.

File: exonum/src/blockchain/mod.rs [hunk 6]
```diff
@@ -338,6 +334,65 @@ impl Blockchain {
...
+            ))?;
+
+            self.tx_from_raw(tx).ok_or(format_err!(
+                "BUG: couldn't create tx from raw message"
```
Comment (DarkEld3r) [inline exonum/src/blockchain/mod.rs]: "BUG" prefix should be removed.

File: exonum/src/blockchain/mod.rs [hunk 7]
```diff
@@ -266,11 +265,8 @@ impl Blockchain {
             let last_hash = self.last_hash();
             // Save & execute transactions
             for (index, hash) in tx_hashes.iter().enumerate() {
-                let tx = pool.get(hash).expect(
-                    "BUG: Cannot find transaction in pool.",
-                );
-
-                execute_transaction(tx.as_ref(), height, index, &mut fork);
+                self.execute_transaction(*hash, height, index, &mut fork)
```
Comment (DarkEld3r) [inline exonum/src/blockchain/mod.rs]: Please add a dot at the end of the comment:
> Save & execute transactions.

File: exonum/src/blockchain/schema.rs [hunk 8]
```diff
@@ -337,19 +344,26 @@ impl<'a> Schema<&'a mut Fork> {
...
         MapIndex::new(TRANSACTIONS, self.view)
     }
 
     /// Mutable reference to the [`transaction_results`][1] index.
     ///
     /// [1]: struct.Schema.html#method.transaction_results
-    pub(crate) fn transaction_results_mut(
-        &mut self,
-    ) -> ProofMapIndex<&mut Fork, Hash, TransactionResult> {
+    pub fn transaction_results_mut(&mut self) -> ProofMapIndex<&mut Fork, Hash, TransactionResult> {
```
Comment (DarkEld3r) [inline exonum/src/blockchain/schema.rs]: `#[doc(hidden)]`?

File: exonum/src/blockchain/tests.rs [hunk 9]
```diff
@@ -14,19 +14,63 @@
...
+use crypto::{gen_keypair, Hash, CryptoHash};
+use storage::{Error, Fork, ListIndex};
+use messages::{Message, RawTransaction};
+use encoding::Error as MessageError;
 use helpers::{Height, ValidatorId};
 
 const IDX_NAME: &'static str = "idx_name";
 
+struct TestService;
+impl Service for TestService {
```
Comment (DarkEld3r) [inline exonum/src/blockchain/tests.rs]: I'm insisting on adding an empty line between `TestService` and its `impl` block.

File: exonum/src/blockchain/tests.rs [hunk 10]
```diff
@@ -14,19 +14,63 @@
...
+    }
+
+    fn tx_from_raw(&self, raw: RawTransaction) -> Result<Box<Transaction>, MessageError> {
+        Ok(Box::new(Tx::from_raw(raw)?))
+    }
+}
+
+transactions! {
+    TestServiceTxs {
+        const SERVICE_ID = 255;
```
Comment (DarkEld3r) [inline exonum/src/blockchain/tests.rs]: Please create the constant for the service id.

File: exonum/src/blockchain/tests.rs [hunk 11]
```diff
@@ -141,64 +185,47 @@ fn gen_tempdir_name() -> String {
...
-            let mut index = ListIndex::new(IDX_NAME, fork);
-            index.push(self.value());
-            index.push(42 / self.value());
-
-            Ok(())
-        }
-    }
+fn insert_tx(schema: &mut Schema<&mut Fork>, hash: Hash, tx: RawTransaction) {
+    schema.transactions_pool_mut().insert(hash);
+    schema.transactions_mut().put(&hash, tx.raw().clone());
```
Comment (DarkEld3r) [inline exonum/src/blockchain/tests.rs]: Why not use `add_transaction_into_pool` here?

Comment (DarkEld3r) [inline exonum/src/blockchain/tests.rs]: Why not use `add_transaction_into_pool` here?

File: exonum/src/blockchain/transaction.rs [hunk 12]
```diff
@@ -808,23 +811,41 @@ mod tests {
...
+            vec![Box::new(TxResultService) as Box<Service>],
+            service_keypair.0,
+            service_keypair.1,
+            ApiSender::new(api_channel.0),
         )
+
+    }
+
+    struct TxResultService;
+    impl Service for TxResultService {
```
Comment (DarkEld3r) [inline exonum/src/blockchain/transaction.rs]: Empty line. :upside_down_face:

File: exonum/src/blockchain/transaction.rs [hunk 13]
```diff
@@ -808,23 +811,41 @@ mod tests {
...
+        fn service_name(&self) -> &'static str {
+            "test service"
+        }
+
+
+        fn state_hash(&self, _: &Snapshot) -> Vec<Hash> {
+            vec![]
+        }
+
+        fn tx_from_raw(&self, raw: RawTransaction) -> Result<Box<Transaction>, ::encoding::Error> {
```
Comment (DarkEld3r) [inline exonum/src/blockchain/transaction.rs]: Please add `use encoding`. :upside_down_face:

File: exonum/src/blockchain/transaction.rs [hunk 14]
```diff
@@ -808,23 +811,41 @@ mod tests {
...
+            Ok(Box::new(TxResult::from_raw(raw)?))
+        }
     }
 
     transactions! {
-        Transactions {
+        TestTx {
```
Comment (DarkEld3r) [inline exonum/src/blockchain/transaction.rs]: We commonly use `tx` as singular form, so it should be `TestTxs`, but I prefer to name it `TestTransactions`.

File: exonum/src/node/consensus.rs [hunk 15]
```diff
@@ -117,29 +117,25 @@ impl NodeHandler {
...
         let snapshot = self.blockchain.snapshot();
-        // Check that transactions are not committed yet
-        for hash in msg.transactions() {
-            if Schema::new(&snapshot).transactions().contains(hash) {
-                error!(
-                    "Received propose with already committed transaction, msg={:?}",
-                    msg
-                );
+        let schema = Schema::new(&*snapshot);
+        //TODO: remove this match after errors refactor.
```
Comment (DarkEld3r) [inline exonum/src/node/consensus.rs]: I propose to remove this "TODO".

File: exonum/src/node/consensus.rs [hunk 16]
```diff
@@ -210,34 +206,42 @@ impl NodeHandler {
...
-                            "Received block with already committed transaction, block={:?}",
-                            msg
-                        );
-                        return;
-                    }
-                    profiler_span!("tx.verify()", {
-                        if !tx.verify() {
-                            error!("Incorrect transaction in block detected, block={:?}", msg);
+            let mut fork = self.blockchain.fork();
+            {
```
Comment (DarkEld3r) [inline exonum/src/node/consensus.rs]: Would it be better to create a separate function instead of nesting scope?

File: exonum/src/node/consensus.rs [hunk 17]
```diff
@@ -259,6 +263,7 @@ impl NodeHandler {
         }
         self.commit(block_hash, msg.precommits().iter(), None);
         self.request_next_block();
+
```
Comment (DarkEld3r) [inline exonum/src/node/consensus.rs]: Please remove this empty line.

File: exonum/src/node/consensus.rs [hunk 18]
```diff
@@ -568,8 +563,17 @@ impl NodeHandler {
                 return;
             }
         });
+        let mut fork = self.blockchain.fork();
+        {
+            let mut schema = Schema::new(&mut fork);
+            schema.transactions_pool_mut().insert(hash);
+            schema.transactions_mut().put(&hash, msg);
```
Comment (DarkEld3r) [inline exonum/src/node/consensus.rs]: Use `add_transaction_into_pool` here?

File: exonum/src/node/consensus.rs [hunk 19]
```diff
@@ -568,8 +563,17 @@ impl NodeHandler {
...
+        }
+        self.blockchain.merge(fork.into_patch()).expect(
+            "Unable to save transaction to persistent pool.",
+        );
```
Comment (DarkEld3r) [inline exonum/src/node/consensus.rs]: As far as I can see, this code is duplicated (thrice :upside_down_face:). It is worth to create a separate function.

File: exonum/src/node/state.rs [hunk 20]
```diff
@@ -757,51 +737,23 @@ impl State {
...
     /// - transaction isn't a part of block
-    pub fn add_transaction(
-        &mut self,
-        tx_hash: Hash,
-        msg: Box<Transaction>,
-        // if tx is in some of propose or in a block,
-        // we should add it, or we could become stuck in some state
-        mut high_priority_tx: bool,
-    ) -> Vec<(Hash, Round)> {
+    pub fn transaction_validated(&mut self, tx_hash: Hash) -> Vec<(Hash, Round)> {
```
Comment (DarkEld3r) [inline exonum/src/node/state.rs]: Is "transactions pool size is exceeded" part still actual?

File: exonum/src/node/state.rs [hunk 21]
```diff
@@ -757,51 +737,23 @@ impl State {
...
 
-    /// Returns non-committed transactions.
-    pub fn transactions(&self) -> &TxPool {
-        &self.transactions
-    }
 
-    /// Adds a transaction to the pool and returns list of proposes that don't contain unknown
-    /// transactions now.
+    /// Handles a transaction validated event and returns list of proposes
+    /// that don't contain unknown transactions now.
```
Comment (DarkEld3r) [inline exonum/src/node/state.rs]: Technically, this function doesn't handle any event, right? Documentation (and probably naming) should be "fixed".

File: exonum/src/node/state.rs [hunk 22]
```diff
@@ -856,23 +808,37 @@ impl State {
...
+        &mut self,
+        msg: &Propose,
+        transactions: &MapIndex<&&Snapshot, Hash, RawMessage>,
+        transaction_pool: &KeySetIndex<&&Snapshot, Hash>,
+    ) -> Result<&ProposeState, ::failure::Error> {
         let propose_hash = msg.hash();
-        let txs = &self.transactions.read().expect("Expected read lock");
         match self.proposes.entry(propose_hash) {
-            Entry::Occupied(..) => None,
+            Entry::Occupied(..) => bail!("Propose already found."),
```
Comment (DarkEld3r) [inline exonum/src/node/state.rs]: I suppose the dot is needless at the end of the the errors description.

File: testkit/src/lib.rs [hunk 23]
```diff
@@ -479,15 +475,17 @@ impl TestKit {
...
         let schema = CoreSchema::new(self.snapshot());
         let uncommitted_txs = transactions.into_iter().filter(|tx| {
-            !schema.transactions().contains(&tx.hash())
+            !schema.transactions().contains(&tx.hash()) ||
+                schema.transactions_pool().contains(&tx.hash())
         });
-
         self.create_block_with_transactions(uncommitted_txs);
         let snapshot = self.snapshot();
+        // rollback commit, and pool update phases.
```
Comment (DarkEld3r) [inline testkit/src/lib.rs]: Please start the comment with a capital letter. :upside_down_face:

File: exonum/src/api/public/system.rs [hunk 24]
```diff
@@ -35,28 +34,26 @@ pub struct HealthCheckInfo {
...
     }
 
     fn mempool_info(self, router: &mut Router) {
         let mempool = move |_: &mut Request| -> IronResult<Response> {
-            let info = MemPoolInfo { size: self.pool.read().expect("Expected read lock").len() };
+            let snapshot = self.blockchain.snapshot();
+            let schema = Schema::new(&snapshot);
+            let pool = schema.transactions_pool();
+            let count = pool.iter().count();
+            let info = MemPoolInfo { size: count };
```
[the remark this answers is not in the document]
Comment (vldm) [reply exonum/src/api/public/system.rs]: i can remove only one line. `let count = pool.iter().count();`

Because `iter()` borrows self, and can't outlive it.

File: exonum/src/node/consensus.rs (hunk 15 above)
Comment (vldm) [reply exonum/src/node/consensus.rs]: `handle_propose` still can't return error 😢

File: exonum/src/node/consensus.rs (hunk 19 above)
Comment (vldm) [reply exonum/src/node/consensus.rs]: Body of `handle_tx` function duplicate `handle_tx_incomming` body. So we should fully refactor this methods (and this will be done in the multithread tx validation). The third duplication is a bit different (it add many tx at once)

File: exonum/src/node/state.rs (hunk 21 above)
Comment (vldm) [reply exonum/src/node/state.rs]: This method is called after transaction validation, how should i rename it (change documentation)?

File: exonum/src/api/public/blockchain_explorer.rs [hunk 25]
```diff
@@ -134,10 +132,21 @@ impl ExplorerApi {
     }
 
     fn transaction_info(&self, hash: &Hash) -> Result<TransactionInfo, ApiError> {
-        if let Some(tx) = self.pool.read().expect("Unable to read pool").get(hash) {
-            Ok(TransactionInfo::InPool {
-                content: tx.serialize_field().map_err(ApiError::InternalError)?,
-            })
+        let snapshot = self.blockchain.snapshot();
+        let schema = Schema::new(&snapshot);
+        if schema.transactions_pool().contains(hash) {
```
Comment (slowli) [inline exonum/src/api/public/blockchain_explorer.rs]: IMO, it could make sense to augment the explorer functionality with mempool-related functions. Can be done later though.

File: exonum/src/blockchain/schema.rs [hunk 26]
```diff
@@ -331,13 +338,22 @@ where
...
+    }
 }
 
 impl<'a> Schema<&'a mut Fork> {
     /// Mutable reference to the [`transactions`][1] index.
     ///
     /// [1]: struct.Schema.html#method.transactions
-    pub(crate) fn transactions_mut(&mut self) -> MapIndex<&mut Fork, Hash, RawMessage> {
+    #[doc(hidden)]
+    pub fn transactions_mut(&mut self) -> MapIndex<&mut Fork, Hash, RawMessage> {
```
Comment (slowli) [inline exonum/src/blockchain/schema.rs]: What's the reasoning for reverting `pub(crate)` visibility? I seriously doubt that there is a legitimate purpose for external apps to modify this index.

File: exonum/src/blockchain/schema.rs [hunk 27]
```diff
@@ -464,4 +488,11 @@ impl<'a> Schema<&'a mut Fork> {
         self.configs_actual_from_mut().push(cfg_ref);
         // TODO: clear storages
     }
+
+    /// Adds transaction into persistent pool
+    #[doc(hidden)]
+    pub fn add_transaction_into_pool(&mut self, tx: RawMessage) {
```
Comment (slowli) [inline exonum/src/blockchain/schema.rs]: Now, this method as opposed to `transactions_mut` and `transactions_pool_mut` can arguably be made public, although, of course, it is still arguably too low-level for most applications.

File: exonum/src/blockchain/transaction.rs [hunk 28]
```diff
@@ -752,17 +754,21 @@ mod tests {
...
-                Height(index),
-                &[transaction.hash()],
-                &pool,
-            );
+            let hash = transaction.hash();
+            {
+                let mut fork = blockchain.fork();
+                {
+                    let mut schema = Schema::new(&mut fork);
+                    schema.transactions_mut().put(
```
Comment (slowli) [inline exonum/src/blockchain/transaction.rs]: Please use `add_transaction_into_pool` here.

File: exonum/src/node/consensus.rs [hunk 29]
```diff
@@ -210,34 +206,40 @@ impl NodeHandler {
...
-                        if !tx.verify() {
-                            error!("Incorrect transaction in block detected, block={:?}", msg);
+            let mut fork = self.blockchain.fork();
+            {
+                let mut schema = Schema::new(&mut fork);
+
+                for raw in msg.transactions() {
+                    if let Some(tx) = self.blockchain.tx_from_raw(raw) {
+                        let hash = tx.hash();
+                        if schema.transactions().get(&hash).is_some() {
```
Comment (slowli) [inline exonum/src/node/consensus.rs]: Probably I'm missing something, but I'm not sure I understand how this check works now. If I understand correctly, transactions in the mempool are in the `transactions()` index, so it may be the case that some of them are uncommitted? Also, could it make sense to use `contains(&hash)` method instead of `get(&hash).is_some()`?

File: exonum/src/node/consensus.rs [hunk 30]
```diff
@@ -579,30 +579,23 @@ impl NodeHandler {
 
     /// Handles external boxed transaction. Additionally transaction will be broadcast to the
     /// Node's peers.
+    #[cfg_attr(feature = "cargo-clippy", allow(needless_pass_by_value))]
     pub fn handle_incoming_tx(&mut self, msg: Box<Transaction>) {
```
Comment (slowli) [inline exonum/src/node/consensus.rs]: This will definitely break the revertible database in the testkit - it assumes that each merge corresponds to a new block. I think a good solution would be to strongly separate non-consensus and consensus storage interfaces (we've discussed it a bit previously).

File: sandbox/src/sandbox.rs [hunk 31]
```diff
@@ -465,29 +465,52 @@ impl Sandbox {
...
+            let mut recover = BTreeSet::new();
+            let mut fork = blockchain.fork();
+            {
+                let mut schema = Schema::new(&mut fork);
+                for raw in txs {
+                    let hash = raw.hash();
+                    hashes.push(hash);
+                    if schema.transactions().get(&hash).is_none() {
+                        recover.insert(hash);
+                        schema.transactions_mut().put(&hash, raw.clone());
```
Comment (slowli) [inline sandbox/src/sandbox.rs]: Please use `schema.add_transaction_into_pool` here.

File: testkit/src/lib.rs [hunk 32]
```diff
@@ -634,29 +624,52 @@ impl TestKit {
...
+        {
+            let blockchain = self.blockchain_mut();
+            let fork = blockchain.fork();
+            blockchain.merge(fork.into_patch()).unwrap();
+        }
         self.do_create_block(&tx_hashes);
     }
 
+    /// Adds transaction into persistent pool.
+    pub fn add_tx(&mut self, transaction: RawMessage) {
```
Comment (slowli) [inline testkit/src/lib.rs]: I don't think that there is a need for such a method; it seems to duplicate `testkit.api().send(transaction)`. (Also, it can probably break some invariants, such as putting non-verifiable or confirmed transactions into the pool; although, admittedly, I haven't analyzed this precisely.)

File: testkit/src/lib.rs [hunk 33]
```diff
@@ -634,29 +624,52 @@ impl TestKit {
...
+        let mut fork = self.blockchain.fork();
+        let mut schema = CoreSchema::new(&mut fork);
+        schema.add_transaction_into_pool(transaction)
+    }
+
+    /// Checks if transaction can be found in pool
+    pub fn is_tx_in_pool(&self, tx_hash: &Hash) -> bool {
```
Comment (slowli) [inline testkit/src/lib.rs]: I'd rather return the (wrapper around?) the readonly pool index, so the user can do more operations with it if he chooses to.

File: testkit/tests/counter/counter.rs [hunk 34]
```diff
@@ -97,6 +96,7 @@ impl Transaction for TxIncrement {
     // This method purposely does not check counter overflow in order to test
     // behavior of panicking transactions.
     fn execute(&self, fork: &mut Fork) -> ExecutionResult {
+        trace!("execute tx {}", self.by());
```
Comment (slowli) [inline testkit/tests/counter/counter.rs]: Are these debugging artifacts? If so, please remove.

File: testkit/tests/service_hooks/main.rs [hunk 35]
```diff
@@ -35,6 +37,10 @@ fn test_handle_commit() {
     for i in 1..5 {
         testkit.create_block();
         let tx = TxAfterCommit::new_with_signature(Height(i), &Signature::zero());
-        assert!(testkit.mempool().contains_key(&tx.hash()));
+        testkit.add_tx(tx.raw().clone());
+        let snapshot = testkit.blockchain_mut().snapshot();
```
Comment (slowli) [inline testkit/tests/service_hooks/main.rs]: There is a `testkit.snapshot()` method.

File: testkit/tests/service_hooks/main.rs [hunk 36]
```diff
@@ -35,6 +37,10 @@ fn test_handle_commit() {
     for i in 1..5 {
         testkit.create_block();
         let tx = TxAfterCommit::new_with_signature(Height(i), &Signature::zero());
-        assert!(testkit.mempool().contains_key(&tx.hash()));
+        testkit.add_tx(tx.raw().clone());
```
Comment (slowli) [inline testkit/tests/service_hooks/main.rs]: Sorry, I don't get it. Isn't this cheating? Afaik, the test should check that `tx` appears in the mempool *without* manually adding it there.

File: testkit/src/lib.rs [hunk 37]
```diff
@@ -465,10 +458,13 @@ impl TestKit {
     /// ```
     pub fn rollback(&mut self, blocks: usize) {
         assert!(
-            (blocks as u64) <= self.height().0,
+            self.height().0 >= blocks as u64,
             "Cannot rollback past genesis block"
         );
-        self.db_handler.rollback(blocks);
+        // Each block contain at least two phases:
```
Comment (slowli) [inline testkit/src/lib.rs]: "At least two" isn't good enough, unfortunately. I think the best way to deal with the problem would be to add a method to the revertible database handler to manually report when new blocks are created (correspondingly, when new checkpoints should be created).

File: exonum/src/api/public/blockchain_explorer.rs (hunk 25 above)
Comment (vldm) [reply exonum/src/api/public/blockchain_explorer.rs]: Can you show an example of "mempool-related functions" that can be add into explorer?

File: exonum/src/blockchain/schema.rs (hunk 26 above)
Comment (vldm) [reply exonum/src/blockchain/schema.rs]: Testkit and sandbox need access to mempool. For now user can create index with the same name as core tables, so we decided to make clever workaround later, after this problem will be fixed. 
Btw this index is hide and undocumented, so user shouldn't touch it.

File: exonum/src/node/consensus.rs (hunk 15 above)
Comment (vitvakatu) [reply exonum/src/node/consensus.rs]: let's mention here related issue

File: exonum/src/node/state.rs (hunk 21 above)
Comment (vitvakatu) [reply exonum/src/node/state.rs]: I think that docs for this method need to be updated/rewritten

File: exonum/src/node/state.rs [hunk 38]
```diff
@@ -856,23 +808,37 @@ impl State {
...
             Entry::Vacant(e) => {
-                let unknown_txs = msg.transactions()
-                    .iter()
-                    .filter(|tx| !txs.contains_key(tx))
-                    .cloned()
-                    .collect::<HashSet<Hash>>();
+                let mut unknown_txs = HashSet::new();
+                for hash in msg.transactions() {
+                    if transactions.get(hash).is_some() {
+                        if !transaction_pool.contains(hash) {
```
Comment (vitvakatu) [inline exonum/src/node/state.rs]: let's merge those two `if`s with `||`, maybe? I don't like the amount of nested scopes

File: testkit/src/lib.rs [hunk 39]
```diff
@@ -596,31 +581,36 @@ impl TestKit {
...
+            let hashes = {
+                let mut schema = CoreSchema::new(&mut fork);
+
+                txs.into_iter()
+                    .filter(|tx| tx.verify())
+                    .map(|tx| {
+                        let tx_id = tx.hash();
+                        assert!(
+                            !schema.transactions().contains(&tx_id) ||
+                                schema.transactions_pool().contains(&tx_id),
```
Comment (vitvakatu) [inline testkit/src/lib.rs]: I think logical not is missing here.... It's hard to understand this piece of code, let's use temporal var?

File: testkit/src/lib.rs [hunk 40]
```diff
@@ -634,29 +624,52 @@ impl TestKit {
...
     pub fn create_block(&mut self) {
         self.poll_events();
 
-        let tx_hashes: Vec<_> = self.mempool().keys().cloned().collect();
-
+        let snapshot = self.blockchain.snapshot();
+        let schema = CoreSchema::new(&snapshot);
+        let txs = schema.transactions_pool();
+        let tx_hashes: Vec<_> = txs.iter().collect();
+        //FIXME: every block should contain two merges
```
Comment (vitvakatu) [inline testkit/src/lib.rs]: IMO `FIXME` comments are suitable for `WIP` status only. Let's replace it with `TODO` and mention relative issue (create one if not yet)

File: exonum/src/blockchain/schema.rs [hunk 41]
```diff
@@ -331,13 +338,22 @@ where
     fn next_height(&self) -> Height {
         Height(self.block_hashes_by_height().len())
     }
+
+    /// Count txs in pool
+    #[cfg_attr(feature = "cargo-clippy", allow(let_and_return))]
+    pub fn pool_len(&self) -> usize {
+        let pool = self.transactions_pool();
+        let count = pool.iter().count();
```
[the remark this answers is not in the document]
Comment (vitvakatu) [reply exonum/src/blockchain/schema.rs]: It's definitely not

File: exonum/src/node/state.rs (hunk 38 above)
Comment (vldm) [reply exonum/src/node/state.rs]: I cant use `or` there because of three possible cases:
- Transaction was found in database but not found in pool - this transaction already commited and propose of this transaction is invalid.
- Transaction was found in database and it is in pool - this transaction is valid for new block.
- Transaction was not found in pool - this transaction is unknown, we need to request it later.

File: exonum/src/blockchain/schema.rs [hunk 42]
```diff
@@ -464,4 +488,11 @@ impl<'a> Schema<&'a mut Fork> {
         self.configs_actual_from_mut().push(cfg_ref);
         // TODO: clear storages
     }
+
+    /// Adds transaction into persistent pool
+    #[doc(hidden)]
+    pub fn add_transaction_into_pool(&mut self, tx: RawMessage) {
+        self.transactions_pool_mut().insert(tx.hash());
+        self.transactions_mut().put(&tx.hash(), tx);
```
[the remark this answers is not in the document]
Comment (vldm) [reply exonum/src/blockchain/schema.rs]: `transactions` schema for now changing its purposes and collect all kind of validated Transactions (commited, inpool)

File: testkit/src/lib.rs (hunk 40 above)
Comment (vldm) [reply testkit/src/lib.rs]: IMO `FIXME`s is other wording `TODO`s.

Comment (vitvakatu) [reply testkit/src/lib.rs]: Fair enough, let's attach issue anyway

File: CHANGELOG.md [hunk 43]
```diff
@@ -5,6 +5,13 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
 
 ## [Unreleased]
 
+### Internal improvements
```
Comment (DarkEld3r) [inline CHANGELOG.md]: Please move 'internal improvements" section below "breaking changes" section.

File: CHANGELOG.md [hunk 44]
```diff
@@ -20,6 +27,17 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
   `blockchain::gen_prefix` method has been removed. Instead, any type that
   implements `StorageKey` trait, can serve as an `index_id`. (#531)
 
+#### exonum-testkit
+
+- Testkit api now contain two methods to work with transaction pool:
+  - `is_tx_in_pool` - for checking transaction existence in pool;
+  - `add_tx` - for adding new transaction into pool.
```
Comment (DarkEld3r) [inline CHANGELOG.md]: for adding *a* new transaction into *the* pool.

File: CHANGELOG.md [hunk 45]
```diff
@@ -20,6 +27,17 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
   `blockchain::gen_prefix` method has been removed. Instead, any type that
   implements `StorageKey` trait, can serve as an `index_id`. (#531)
 
+#### exonum-testkit
+
+- Testkit api now contain two methods to work with transaction pool:
```
Comment (DarkEld3r) [inline CHANGELOG.md]: ...*contains* ...with *the* transaction pool

File: CHANGELOG.md [hunk 46]
```diff
@@ -20,6 +27,17 @@ The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
...
+  - `is_tx_in_pool` - for checking transaction existence in pool;
```
Comment (DarkEld3r) [inline CHANGELOG.md]: in *the* pool

File: exonum/benches/block.rs [hunk 47]
```diff
@@ -23,37 +23,62 @@ extern crate test;
...
 
+            fn state_hash(&self, _: &Snapshot) -> Vec<Hash> {
+                Vec::new()
+            }
+
+            fn tx_from_raw(&self, raw: RawTransaction) -> Result<Box<Transaction>, EncodingError> {
+                Ok(Box::new(Tx::from_raw(raw)?))
+            }
+        }
+        transactions! {
```
Comment (DarkEld3r) [inline exonum/benches/block.rs]: I would add an empty line before `transactions!` too. :upside_down_face:

File: exonum/benches/block.rs [hunk 48]
```diff
@@ -23,37 +23,62 @@ extern crate test;
...
     use exonum::crypto::{gen_keypair, CryptoHash, Hash, PublicKey, SecretKey};
-    use exonum::messages::Message;
+    use exonum::messages::{Message, RawTransaction};
+    use exonum::encoding::Error as EncodingError;
     use exonum::helpers::{Height, ValidatorId};
     use exonum::node::ApiSender;
 
-    fn create_blockchain(db: Box<Database>) -> Blockchain {
+    const TIMESTAMPING_SERVICE_ID: u16 = 1;
+    const CRYPTOCURRENCY_SERVICE_ID: u16 = 1;
```
Comment (DarkEld3r) [inline exonum/benches/block.rs]: Please use a different constant or one `SERVICE_ID` constant for both.

File: exonum/src/blockchain/mod.rs [hunk 49]
```diff
@@ -256,24 +257,22 @@ impl Blockchain {
...
+            // Save & execute transactions.
             for (index, hash) in tx_hashes.iter().enumerate() {
-                let tx = pool.get(hash).expect(
-                    "BUG: Cannot find transaction in pool.",
-                );
-
-                execute_transaction(tx.as_ref(), height, index, &mut fork);
+                self.execute_transaction(*hash, height, index, &mut fork)
+                    // execute_transaction could fail,with invalid
+                    // Transaction that cannot be deserialized, or not found in pool.
```
Comment (DarkEld3r) [inline exonum/src/blockchain/mod.rs]: Execution could fail if the transaction cannot be deserialized or it isn't found.

File: exonum/src/blockchain/schema.rs [hunk 50]
```diff
@@ -331,6 +338,15 @@ where
     fn next_height(&self) -> Height {
         Height(self.block_hashes_by_height().len())
     }
+
+    /// Count txs in pool
```
Comment (DarkEld3r) [inline exonum/src/blockchain/schema.rs]: "Counts transactions in the pool" or "returns number of transactions in the pool".

File: exonum/src/blockchain/schema.rs [hunk 51]
```diff
@@ -331,6 +338,15 @@ where
...
+    #[cfg_attr(feature = "cargo-clippy", allow(let_and_return))]
+    pub fn pool_len(&self) -> usize {
```
Comment (DarkEld3r) [inline exonum/src/blockchain/schema.rs]: `tx_pool_len`?

File: exonum/src/blockchain/schema.rs [hunk 52]
```diff
@@ -331,6 +338,15 @@ where
...
+        let pool = self.transactions_pool();
+        // TODO: Change count to other method with O(1) complexity (ECR-977)
```
Comment (DarkEld3r) [inline exonum/src/blockchain/schema.rs]: Dot at the end. :upside_down_face:

File: exonum/src/node/consensus.rs [hunk 53]
```diff
@@ -487,13 +489,11 @@ impl NodeHandler {
             );
             (block_state.txs().len(), block_state.proposer_id())
         };
+        let snapshot = self.blockchain.snapshot();
+        let schema = Schema::new(&snapshot);
+        let pool = schema.pool_len();
```
Comment (DarkEld3r) [inline exonum/src/node/consensus.rs]: `pool_size` or `pool_len`?

File: exonum/src/blockchain/mod.rs [hunk 54]
```diff
@@ -256,24 +257,22 @@ impl Blockchain {
...
+            // Save & execute transactions.
             for (index, hash) in tx_hashes.iter().enumerate() {
-                let tx = pool.get(hash).expect(
-                    "BUG: Cannot find transaction in pool.",
-                );
-
-                execute_transaction(tx.as_ref(), height, index, &mut fork);
+                self.execute_transaction(*hash, height, index, &mut fork)
+                    // Execution could fail if the transaction
+                    // cannot be deserialized or it isn't in pool.
```
Comment (DarkEld3r) [inline exonum/src/blockchain/mod.rs]: ...in the pool. :upside_down_face:

File: exonum/src/node/consensus.rs [hunk 55]
```diff
@@ -155,6 +151,40 @@ impl NodeHandler {
         }
     }
 
+    // Validates transaction from block, and append it into pool.
+    // Returns tuple of transactions hashes and Patch of changes in pool.
```
Comment (DarkEld3r) [inline exonum/src/node/consensus.rs]: ...transaction hashes...in the pool.

File: exonum/src/node/consensus.rs [hunk 56]
```diff
@@ -155,6 +151,40 @@ impl NodeHandler {
         }
     }
 
+    // Validates transaction from block, and append it into pool.
```
Comment (DarkEld3r) [inline exonum/src/node/consensus.rs]: ...from the block and append**s** them into the pool.
